actors/miner/tests/util.rs

```diff
@@ -1682,12 +1684,452 @@ pub struct DeadlineStateSummary {
+    fn create<BS: Blockstore>(
+        expiration_queue: &ExpirationQueue<BS>,
+        live_sectors: &SectorsMap,
+        partition_faults: &BitField,
+        quant: QuantSpec,
+        sector_size: SectorSize,
+        acc: &mut MessageAccumulator,
+    ) -> Self {
+        let mut seen_sectors: HashSet<SectorNumber> = HashSet::new();
+        let mut all_on_time: Vec<BitField> = Vec::new();
+        let mut all_early: Vec<BitField> = Vec::new();
+        let mut expiration_epochs: Vec<ChainEpoch> = Vec::new();
+        let mut all_active_power = PowerPair::zero();
+        let mut all_faulty_power = PowerPair::zero();
+        let mut all_on_time_pledge = BigInt::zero();
+        let mut first_queue_epoch: ChainEpoch = -1;
+
+        let ret = expiration_queue.amt.for_each(|epoch, expiration_set| {
+            let epoch = epoch as i64;
+            let mut acc = acc.with_prefix(&format!("expiration epoch {epoch}: "));
+            let quant_up = quant.quantize_up(epoch);
+            acc.require(quant_up == epoch, &format!("expiration queue key {epoch} is not quantized, expected {quant_up}"));
+            if first_queue_epoch == -1 {
+                first_queue_epoch = epoch;
+            }
+
+            expiration_epochs.push(epoch);
+
+            let mut on_time_sectors_pledge = BigInt::zero();
+            for sector_number in expiration_set.on_time_sectors.iter() {
+                // check sectors are present only once
+                if !seen_sectors.insert(sector_number) {
+                    acc.add(&format!("sector {sector_number} in expiration queue twice"));
+                }
+
+                // check expiring sectors are still alive
+                if let Some(sector) = live_sectors.get(&sector_number) {
+                    // The sector can be "on time" either at its target expiration epoch, or in the first queue entry
+                    // (a CC-replaced sector moved forward).
+                    let target = quant.quantize_up(sector.expiration);
+                    acc.require(epoch == target || epoch == first_queue_epoch, &format!("invalid expiration {epoch} for sector {sector_number}, expected {first_queue_epoch} or {target}"));
+                    on_time_sectors_pledge += sector.initial_pledge.clone();
+                } else {
+                    acc.add(&format!("on time expiration sector {sector_number} isn't live"));
+                }
+            }
+
+            for sector_number in expiration_set.early_sectors.iter() {
+                // check sectors are present only once
+                if !seen_sectors.insert(sector_number) {
+                    acc.add(&format!("sector {sector_number} in expiration queue twice"));
+                }
+
+                // check early sectors are faulty
+                acc.require(partition_faults.is_empty() || partition_faults.get(sector_number), &format!("sector {sector_number} expiring early but not faulty"));
```

Comment:
I think `partitionFaultsMap == nil` from go is a bug.  I think it was intending to be `!= nil` to avoid panicing.  If we keep it this way then when we have an expiration set with an early termination but a partition with no faults then the state invariant checks won't see an error.  But this state is erroneous as it breaks the invariant that all early sectors are faulty.

Reply:
Sounds good, should it be also fixed on Go side?

Reply:
The Go code is correct, but this one is a bug.

It's correct in Go because `nil` != empty, and the nil is being used to signal that the data is not available because it could not be read (in which case an error will also already have been accumulated). That's because the partition faults map is loaded inside this function. It does rely on the somewhat shaky assumption that `partitionFaults.AllMap` will return an empty map rather than nil if there are no faults. This was introduced in https://github.com/filecoin-project/specs-actors/pull/1242.

In this code the partition faults map is being passed as a parameter (why) and can't be "missing".

Reply:
Thanks for checking it out, @anorth. Do you mean the code commented by @ZenGround0 is buggy? Is it correct? Not the one that was ultimately merged (i.e. without the `is_empty` check)?